perf: use size prefixing for zstd buffer compressor for better decompressing performance

niyue opened the pull request:
# Summary

This PR addresses [lancedb/lance#4028](https://github.com/lancedb/lance/issues/4028) by introducing a length-prefixed format for the Zstd buffer compressor. The goal is to improve the decompression performance of Zstd-compressed buffers.

# Solution

Currently, the Zstd buffer compressor uses the standard compression API and decompresses data using the Zstd streaming API. While functional, this approach is not optimal as the streaming decompression introduces overhead.

In this PR, I switch to a block decompression approach by:
* Prefixing each compressed buffer with its original (uncompressed) length (encoded as a 64-bit little-endian integer)
* Using Zstd’s block decompression API during decoding, which avoids the overhead of stream mode.
* Leveraging the Zstd magic number and a reserved bit in the Zstd frame header descriptor (as defined in RFC 8878) to distinguish between a standard Zstd stream and the new length-prefixed format, so that compatibility can be guaranteed 

# Consistency with Existing Approaches
* This technique is consistent with the one used by the [Apache Arrow IPC format][1][2], which also embeds the uncompressed size to enable faster, more efficient decompression.
* The LZ4 buffer compressor in lance already follows a similar approach, as its API natively supports it via the `prepend_size` parameter in the `compress_to_buffer` function.

# Benchmark Results

In limited benchmarks across several input sizes:

* Decompression speed improved by 30% to 200%, depending on the data.
* Compression speed remains unchanged.
* Compressed size increases by 8 bytes (due to the uncompressed size prefix).

# References
* [1] Arrow IPC writer: https://github.com/apache/arrow/blob/bd31f83aaa93db8427bc90285658e29d79ce5efd/cpp/src/arrow/ipc/writer.cc#L230-L233
* [2] Arrow IPC reader: https://github.com/apache/arrow/blob/bd31f83aaa93db8427bc90285658e29d79ce5efd/cpp/src/arrow/ipc/reader.cc#L398-L407

niyue commented on rust/lance-encoding/src/encodings/physical/block_compress.rs:
```diff
@@ -103,8 +103,23 @@ impl BufferCompressor for ZstdBufferCompressor {
-        let source = Cursor::new(input_buf);
-        zstd::stream::copy_decode(source, output_buf)?;
+        // read the first 64-bit prefix as the uncompressed buffer length
+        let mut input_buf = input_buf;
+        let mut uncompressed_len_buf = [0u8; 8];
+        uncompressed_len_buf.copy_from_slice(&input_buf[..8]);
+        let uncompressed_len = u64::from_le_bytes(uncompressed_len_buf) as usize;
+        input_buf = &input_buf[8..];
+
+        let input_len = input_buf.len();
+        // resize output buffer to the uncompressed length
+        output_buf.resize(uncompressed_len, 0);
+        let output_len = output_buf.len();
+        let result = decompress_to_buffer(input_buf, output_buf);
+        if let Err(e) = result {
+            error!("Failed to decompress zstd buffer: input_buffer size: {}, output_buffer size: {}, error: {:?}", input_len, output_len, e);
+            return Err(e.into());
+        }
+
```
Okay, let me explore if there's a better way to handle compatibility.

niyue commented on rust/lance-encoding/src/encodings/logical/primitive.rs:
```diff
@@ -3346,7 +3346,15 @@ impl PrimitiveFieldEncoder {
                 row_number: 0, // legacy encoders do not use
             })
         })
-        .map(|res_res| res_res.unwrap())
```
This is probably less relevant with this PR, but previously the `unwrap` here causing the program to crash since the encoding task error is not handled

westonpace commented on rust/lance-encoding/src/encodings/logical/primitive.rs:
```diff
@@ -3346,7 +3346,15 @@ impl PrimitiveFieldEncoder {
-        .map(|res_res| res_res.unwrap())
+        .map(|res_res| {
+            res_res.unwrap_or_else(|err| {
+                log::error!("Encoding task failed with error: {:?}", err);
+                Err(lance_core::Error::Internal {
```
We don't generally log _and_ raise but it's probably fine here (this should be relatively rare?)

westonpace commented on rust/lance-encoding/src/encodings/physical/block_compress.rs:
```diff
@@ -103,8 +156,18 @@ impl BufferCompressor for ZstdBufferCompressor {
-        let source = Cursor::new(input_buf);
-        zstd::stream::copy_decode(source, output_buf)?;
+        if input_buf.is_empty() {
+            output_buf.clear();
```
I don't think we should call `clear` here.  We use this path in 2.1 to decompress multiple values (and each one should add to the same `output_buf` so then a later empty value might clear an earlier decompressed value)

westonpace commented on rust/lance-encoding/src/encodings/physical/block_compress.rs:
```diff
@@ -90,11 +89,65 @@ impl ZstdBufferCompressor {
+        let uncompressed_len = u64::from_le_bytes(len_buf) as usize;
+
+        let compressed_data = &input_buf[8..];
+        output_buf.resize(uncompressed_len, 0);
```
Can you use `reserve` instead of `resize`?  Since this could be called multiple times with the same output buf.

niyue replied:
@westonpace 
Since zstd's `decompress_to_buffer` requires the output_buffer to be pre-allocated, I still use the `resize`.
However, I’ve added a `start` offset to ensure that this function (and the `compress` function) can be called multiple times with the same output buffer. A dedicated test case has been added to verify this usage.

westonpace reviewed (changes requested):
I have a few suggestions to avoid breaking the 2.1 path here: https://github.com/lancedb/lance/blob/a499cfa06b7221b895bc13908cfc2ee7aadba46e/rust/lance-encoding/src/encodings/physical/block_compress.rs#L267

In that path we first allocate a buffer and then call `decompress` multiple times with the same output buffer.

westonpace reviewed (approved):
This will cause potential forward compatibility issues.  Files written with newer versions of lance will not be readable in older versions of lance.  This is not a concern for us (lancedb) as we don't make use of this feature when using 2.0.

@yanghua do you have any concerns?  If not, I think we can merge once CI passes.

yanghua reviewed (approved):
@niyue Thanks for improving the performance and the detailed explanation.

niyue commented:
There was a significant refactoring on the main branch, and the two files I modified have been renamed or moved. I've updated them based on the latest main branch. Please let me know if there's anything that doesn't look right.

niyue commented:
Hey @westonpace, just a quick ping on this PR. CI is green and please let me know if there’s anything else needed, or if it’s good to merge. Appreciate your time!
